pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+
+    .. note::
+        While the Fourier spectrum usually does not depend
+        on trainable circuit parameters or the actual values of the inputs,
+        it may still change based on inputs to the QNode that alter the architecture
+        of the circuit.
+
+    Above, we selected all input-encoding parameters for the spectrum computation, using
+    the ``argnum`` keyword argument. We may also restrict the full analysis to a single
+    QNode argument, again using ``argnum``:
+
+    >>> res = advanced_spectrum(circuit, argnum=[0])(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "x": {(0,): [-0.5, 0.0, 0.5], (1,): [-0.5, 0.0, 0.5], (2,): [-0.5, 0.0, 0.5]}
+
+    Selecting arguments by name instead of index is possible via the
+    ``encoding_args`` argument:
+
+    >>> res = advanced_spectrum(circuit, encoding_args={"y"})(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "y": {(0,): [-2.3, 0.0, 2.3], (1,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+
+    Note that for array-valued arguments the spectrum for each element of the array
+    is computed. A more fine-grained control is available by passing index tuples
+    for the respective argument name in ``encoding_args``:
+
+    >>> encoding_args = {"y": [(0,),(2,)]}
+    >>> res = advanced_spectrum(circuit, encoding_args=encoding_args)(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "y": {(0,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+
+    .. warning::
+        The ``advanced_spectrum`` function does not check if the result of the
+        circuit is an expectation value. It checks whether the classical preprocessing between
+        QNode and gate arguments is linear by computing the Jacobian of the processing
+        at multiple points. This makes it unlikely -- *but not impossible* -- that
+        non-linear functions go undetected.
+        The number of additional points at which the Jacobian is computed can be controlled
+        via ``num_pos``, and the test is deactivated if ``num_pos=0`` (discouraged).
+        Furthermore, the QNode arguments *not* marked in ``argnum`` will not be
+        considered in this test and if they resemble encoded inputs, the entire
+        spectrum might be incorrect or the circuit might not even admit one.
+
+    The ``advanced_spectrum`` function works in all interfaces:
+
+    .. code-block:: python
+
+        import tensorflow as tf
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev, interface='tf')
+        def circuit(x):
+            qml.RX(0.4*x[0], wires=0)
+            qml.PhaseShift(x[1]*np.pi, wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = tf.constant([1., 2.])
+        res = advanced_spectrum(circuit)(x)
+
+    >>> print(res)
+    {"x": {(0,): [-0.4, 0.0, 0.4], (1,): [-3.14159, 0.0, 3.14159]}}
```

Comment:
Nice!

only comment is that I would expect
```python
```python
    >>> print(res)
    {"x": {(0,): tf.tensor([-0.4, 0.0, 0.4]), (1,): tf.tensor([-3.14159, 0.0, 3.14159])}}
```

as the output. that is, the `advanced_spectrum` function should ideally not affect the dtype.

Reply:
This would also allow for differentiability!